### ISSUE TITLE
What to name Chapel methods for (bitwise) conversions between values of different types? (e.g., reals and uints)

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#21774]
Issue #14140 is a longstanding request for the ability to convert between floats and uints by re-interpreting their binary representations rather than their values.  Thus, where `3.1415:uint` would give `3`, this capability would look at the bit pattern for `3.1415`, interpret that as a `uint` and return that value.  I have an evolving implementation of this in #21040 and its backing branch, but the user-facing interface for this hasn't received any design attention.  

[Edit: An implementation using the method name `.transmute()` was added in #21813 but I'm leaving this issue open since it did not receive much design discussion or attention outside of Michael, Damian, and myself.  Also, it is currently undocumented and marked as unstable, both of which should be addressed]

Some options include:

Method-based approaches
===================

We could add methods to the real/uint types themselves that support this directly.

toUint/toReal
-------------
I'm putting this here as a straw-person proposal, though I don't think it's a good one because it reads too much like a cast, as though it's the value being preserved

```chapel
var newUint = myReal.toUint();
var newReal = newUint.toReal();  
```

asUint/asReal
-------------
While just a minor tweak to the above, the use of `as` at least suggests a re-interpretation.  But the concern may still hold...

```chapel
var newUint = myReal.asUint();
var newReal = newUint.asReal();  
```

toBits/toFloat
-------------

These names are arguably a slight improvement in that they focus more on the representation than the resulting type—of course, this also means that the resulting type isn't obvious from the usage.

```chapel
var newUint = myReal.toBits();
var newReal = newUint.toFloat();
```

Note also that if `toFloat()` returns a `real`, then there's no way to get an imaginary unless we had it take an optional type argument, like:

```chapel
var newImag = newUint.toFloat(imag);`
```

transmute
----------
Taking a cue from Rust, we could support a `transmute` that took a type argument, like:

```chapel
var newUint = myReal.transmute(uint);
var newReal = newUint.transmute(real);
var newImag = newUint.transmute(imag);
```

This has the advantage of not breaking type names or representations into the method name, though it's likely that only certain patterns will be supported in the near-term.  That said, it arguably provides a general way of supporting other cases in the future.

bitsAs / interpretBitsAs
-----------------------
These are just potentially different names for transmute:

```chapel
var newUint = myReal.bitsAs(uint);
var newReal = newUint.bitsAs(real);
var newImag = newUint.bitsAs(imag);
```

```chapel
var newUint = myReal.interpretBitsAs(uint);
var newReal = newUint.interpretBitsAs(real);
var newImag = newUint.interpretBitsAs(imag);
```

Standalone approaches
=================

Most of the above approaches could also be written as standalone routines, such as:

```chapel
var newUint = toBits(myReal);
var newUint = transmute(myReal, uint);
// etc.
```

**Comment by bradcray** [comment]
Tagging @bryantlam, @damianmoz, and @mppf for their input, as the most active participants on https://github.com/chapel-lang/chapel/issues/14140.

**Comment by mppf** [comment]
FWIW, I like `toBits`, `bitsAs`, and `transmute`. I don't much like `toFloat` `toUint` `toReal` `asUint` `asReal` because these are not using different terminology than I would use to describe something like `3.1415:uint`.

**Comment by damianmoz** [comment]
I will add another two routine pairings, one in the flavour of Rust but at a higher lever than _transmute_
```chapel
var myReal754Encoding = myImage.toBits;
var myNewReal = myReal754Encoding.fromBits;

var myReal754Encoding = myReal.ieee754; // return the ieee754 encoding
var myNewReal = myReal754Encoding.reify; // 'reify' means 'to make something real (or concrete)

assert(myReal == myNewReal);
```
I have a preference for neither.

Briefly, maybe poorly, paraphrasing the IEEE 754 standard, routines mentioned above such as `toBits` when used as a method applied to a **real**(w) datum returns the unique bit-wise encoding of that data in _w_ bits. That is, there is only one unique type that `toBits` will return;

The IEEE 753 standard defines neither arbitrary complex data nor one where the real component is zero, i.e. Chapel's **imag**(w) type. So from an IEEE 754 perspective, working with **imag**(w) types is undefined. 

But if we really want to handle an **imag**(w) type, we can define the methods with names like `toBits` or `ieee754` to suit outselves. So
```chapel
var myImag754Encoding = myImag.toBits;
var myImage754Encoding = myImag.ieee754;
```
seems reasonable (even thought there is no such thing as an imaginary encoding). But then, because of that same fact, i.e. that there is no such as an imaginary encoding, only the following (mouthful) operation makes sense from an IEEE 754 perspective, then one can only really say that
```chapel
var myNewImag = (myImag754Encoding.fromBits):imag(numBits(myImag754Encoding));

assert(myNewImag == myImag);
```
And yes, in this last case of transmutation back to an **imag**(w) type, you could pass to the method, say `fromBits`, the type **imag**(w), but that assumes that you check at compile time within that method that
```chapel
compilerAssert(numBits(this) == w);
```
I will let people comment on my ramblings before saying anything else.  Or maybe somebody else wants to comment further.

**Comment by bradcray** [comment]
Thanks for the input, Damian and Michael.

Even though IEEE 754 might not say anything about `imag`, I don't see any reason to preclude it from a language standpoint, and would be more inclined to support it via something like `.transmute(imag)` / `.fromBits(imag)` rather than casting a transmuted/reified real into an imag, in part for cleanliness in the interface, and in part because the real->imag cast essentially seems to be saying "ignore the mathematical value and just interpret the bits as this different type" which is pretty much what this feature is doing anyway—so why require the user to use two similar-but-differently-spelled operations to get to the type they want rather than just one.

The main downside to that I can think of is that it would either require saying `.fromBits(real)` / `.transmute(real)` for the common case of going from bits to real rather than, say, just `.fromBits()`.  Or we could add a default `type = real(w)` argument, but that seems potentially too real-centric (if, say, we ultimately allowed any w-bit type to be transmuted to any other through these routines).

For my own tastes, I'm leaning more toward `.transmute(type t)` than most of the other preferred suggestions because I like that it phrases the operation independent of source/target type; and also because I'm feeling a little bit nervous about equating `uint` with being 'a bunch of bits [with no other mathematical meaning]', particularly anticipating a future in which we might have a `bits` or 'bit vector' type of some sort.  That is, I think `uint` is the most reasonable type we have to convert `real`s into at present, but I'm worried that saying `myReal.toBits()` doesn't obviously say "make it a uint", whereas `myReal.transmute(uint)` does.  The main downside to this approach that I see is that in the short-term, we're likely to only support certain combinations of types, so what looks like a completely general interface will be somewhat sparse in its support.  That said, I also like that it opens the door to supporting other transmutations in the future.

Thoughts?  And Damian, did you have more you wanted to say that you were holding back for now?

(Note that however this discussion evolves in the next few days, my hope is to have a draft of this in 1.30 that is usable, but that will be marked as unstable until we have a proper and broader discussion about naming).

**Comment by damianmoz** [comment]
My earlier thoughts  were more to throw a few more ideas into the ring. Some of my thoughts could be total rubbish.

While my interest is to go to and from the IEEE 754 encoding, is the question about the whole **transmutation** concept which is a bigger issue far beyond floating point issues. Are there any other questions which need to be raised? I worry my field of vision on this is too limited.

Forgive my ramblings that follow. I have agonized on the following text for a few hours and still do not like it. But I need to go and do some other work so I will put it out there:

I am all for cleanliness of interface.  But in the context of floating point transmutation only, something in the back of my head says to stay away from **imag**(w). I cannot really think why. Maybe I am being too rigorous about the concepts in the standard. Should you also include **complex**(w) types in the mix? But then, that seems like a massive complication. Mind you, if you have a general **transmutation** feature, then going to and from **imag**.(w) is perfectly legitimate. Is transmutation allowed on all basic types?

**Comment by damianmoz** [comment]
I agree with you. I was just trying to flag more discussion points just in case we, or more specifically I, had missed something.

In that case, the generalized **transmute** approach is the way to go, even if there is yet to be agreement on that word.

**Comment by damianmoz** [comment]
In deference to Americans, I referred to the Merriam-Webster dictionary and `transmute` is about the best word in English for the task. Somebody already went through this exercise anyway in Rust and got it right.

Of its synonyms, the word `recast` is also a candidate but I think it has too many alternate connotations given its past usage in C++, the word `transmogrify` is too esoteric and implies a sense of physical change, and `convert` is far too broad. And the rest do not really cut it.

At nine letters, `transmute` is a bit long but in the context in which it will be used in my own usage, will not impact readability. as far as I can tell.

**Comment by bradcray** [comment]
That's funny that you mention "transmogrify", as I had considered proposing it (jokingly) as a tribute to Calvin and Hobbes:

![image](https://user-images.githubusercontent.com/7536222/223910391-752e5d38-f54c-4d1d-b014-0582acfc17b0.png)

**Comment by damianmoz** [comment]
I am saying very little here except that my education must be sorely/seriously/massively lacking because those two names do not ring bells for an Australian, or at least this Australian. I am totally lost for what sort of emoji to use in response.

I will quote Merriam-Webster and say that " transmogrify suggests a strange or preposterous metamorphosis" which is not what we are about when we want to look at the bits, We want no change per se on the original datum, we are just exposing the bit pattern which is it in memory representation  to (deep) introspection.

**Comment by bradcray** [comment]
My current transmute-based implementation is here:  https://github.com/chapel-lang/chapel/issues/21774

**Comment by bradcray** [comment]
I've now merged the transmute-based implementation in #21813, but am leaving this issue open since it did not receive much design attention outside of Michael, Damian, and myself.  The feature should also be documented and the @unstable removed once we're committed to the interface.

**Comment by bradcray** [comment]
I have not heard any negative reviews of the name .transmute() and there is precedent for it, so I'm closing this.
